### PR TITLE
Limit memory usage based on cgroup info

### PR DESCRIPTION
**Pull request by VannTen**
## Related Issues and Dependencies
This should help diagnoses thoth-station/thoth-application#2690
We should get some stacktrace, which will help us locate where our python code allocates a lot. (Especially if we have a space complexity bug/problem)

## This introduces a breaking change

- [ ] Yes
- [x] No

## Description

We try to convert OOM-kill at the cgroup level into python MemoryError.

The method employed has a number of caveats:
- setrlimit only acts on one process, when cgroup is a group of process
  ; other process in the container could also consume memory
- the OOM-killer might still kick in before Python raising MemoryError,
  depending on the particular of the Python allocator

### CODE REVIEW

**Comment by VannTen** [comment]
@goern @harshad16 you might be interested.

File: thoth/solver/cli.py [hunk 1]
```diff
@@ -67,6 +69,30 @@ def cli(ctx=None, verbose=0):
...
+        '/sys/fs/cgroup/memory/memory.limit_in_bytes',
+        '/sys/fs/cgroup/memory.high', #  cgroups v2
+        '/sys/fs/cgroup/memory.max',
+    ]:
+        try:
+            with open(file) as limit:
+                memory = limit.read()
+                if memory == "max": # memory.high might contain "max" -> indicates to use the memory.max value
+                    continue
+                value = int(limit.read())
```
Comment (codificat) [inline thoth/solver/cli.py:88]: isn't this meant to be:
```suggestion
                value = int(memory)
```

**Review by codificat — COMMENTED** [review]
Just one question/comment, but
/approve
the proposal, good idea, thanks

File: thoth/solver/cli.py (hunk 1 above)
Comment (VannTen) [reply thoth/solver/cli.py:88]: Yeah, spot on, thanks :+1:

**Pull request merged**